pandas/_libs/tslibs/timestamps.pyx

```diff
@@ -319,10 +319,10 @@ cdef class _Timestamp(ABCTimestamp):
             else:
                 self = type(other)(self)
 
-            # validate tz's
-            if not tz_compare(self.tzinfo, other.tzinfo):
-                raise TypeError("Timestamp subtraction must have the "
-                                "same timezones or no timezones")
+            if (self.tzinfo is None) ^ (other.tzinfo is None):
+                raise TypeError(
+                    "Cannot compare tz-naive and tz-aware datetime-like objects."
```

Comment:
should this message be about subtraction?

Reply:
Yup, makes sense! just so you know though, it won't be consistent across the board because in some of the vector implementations (e.g. `pandas/core/arrays/datetimes.py`) there is a timezone awareness check which is optimized, and when the check fails there is no way to know what operation was being done. so, some of the array / series based tests i've added will still assert 'cannot _compare_'